storage/rocksdb/ha_rocksdb.cc

```diff
@@ -15634,6 +15672,530 @@ void Rdb_compaction_stats::record_end(rocksdb::CompactionJobInfo info) {
+        // BKA's variant of "Index Condition Pushdown" check failed
+        continue;
+      }
+
+      if (self->mrr_funcs.skip_record &&
+          self->mrr_funcs.skip_record(self->mrr_iter, *range_ptr,
+                                      (uchar*)self->m_last_rowkey.ptr())) {
+        continue;
+      }
+
+      memcpy(buf, self->m_last_rowkey.ptr(), self->m_last_rowkey.length());
+      *size = self->m_last_rowkey.length();
+      break;
+    }
+    return err;
+  }
+
+  bool eof() override {
+    return err != 0;
+  }
+};
+
+// Initialize an MRR scan
+int ha_rocksdb::multi_range_read_init(RANGE_SEQ_IF *seq, void *seq_init_param,
+                                      uint n_ranges, uint mode,
+                                      HANDLER_BUFFER *buf) {
+  int res;
+
+  if (!current_thd->optimizer_switch_flag(OPTIMIZER_SWITCH_MRR) ||
+      (mode & HA_MRR_USE_DEFAULT_IMPL) ||
+      (buf->buffer_end - buf->buffer < mrr_get_length_per_rec()) ||
+      (active_index != table->s->primary_key && (mode & HA_MRR_SORTED))) {
+    mrr_uses_default_impl = true;
+    res = handler::multi_range_read_init(seq, seq_init_param, n_ranges, mode,
+                                         buf);
+    return res;
+  }
+
+  // Ok, using a non-default MRR implementation, MultiGet-MRR
+
+  mrr_uses_default_impl = false;
+  mrr_n_elements = 0;  // nothing to cleanup, yet.
+  mrr_enabled_keyread = false;
+  mrr_rowid_reader = nullptr;
+
+  mrr_funcs = *seq;
+  mrr_buf = *buf;
+
+  bool is_mrr_assoc = !MY_TEST(mode & HA_MRR_NO_ASSOCIATION);
+  if (is_mrr_assoc) {
+    ++table->in_use->status_var.ha_multi_range_read_init_count;
+  }
+
+  mrr_sorted_mode = (mode & HA_MRR_SORTED) != 0;
+
+  if (active_index == table->s->primary_key) {
+    // ICP is not supported for PK, so we don't expect that BKA's variant
+    // of ICP would be used:
+    DBUG_ASSERT(!mrr_funcs.skip_index_tuple);
+    mrr_used_cpk = true;
+    mrr_rowid_reader =
+      new Mrr_pk_scan_rowid_source(this, seq_init_param, n_ranges, mode);
+  } else {
+    mrr_used_cpk = false;
+    auto reader = new Mrr_sec_key_rowid_source(this);
+    reader->init(seq, seq_init_param, n_ranges, mode);
+    mrr_rowid_reader = reader;
+  }
+
+  res = mrr_fill_buffer();
+
+  // note: here, we must NOT return HA_ERR_END_OF_FILE even if we know there
+  // are no matches. We should return 0 here and return HA_ERR_END_OF_FILE
+  // from the first multi_range_read_next() call.
+  if (res == HA_ERR_END_OF_FILE) {
+    res = 0;
+  }
+
+  return res;
+}
+
+// Return the amount of buffer space that MRR scan requires for each record
+// returned
+uint ha_rocksdb::mrr_get_length_per_rec() {
+  return sizeof(rocksdb::Slice) + sizeof(rocksdb::Status) +
+         sizeof(rocksdb::PinnableSlice) +
+         sizeof(char *) +  // this for KEY_MULTI_RANGE::ptr
+         m_pk_descr->max_storage_fmt_length();
+}
+
+template <typename T> void align_ptr(char **p) {
+  if (((size_t)p) % alignof(T)) {
```

Comment:
It looks like there is a bug here we are trying to align pointer to pointer (`p`) instead of pointer (`*p`)
```suggestion
  if ((reinterpret_cast<std::uintptr_t>(*p)) % alignof(T)) {
```